### PR TITLE
Upgrade ingestion to allow for in-flight updates to feature sets for sinks

### PR DESCRIPTION
**Pull request by pyalex**
**What this PR does / why we need it**:

There are several improvements:

* Sink API accepts `PCollection<FeatureSetSpec>` and both BQ and redis can change their processing according to new specs in-flight
* BQ sink writes through file loads (to gs) - that enables schema update & also might be less expensive
* BQ sink produces sucessful inserts stream - enables metrics on this sink
* Dataflow JobManager stop+start job when specs is updated. This is caused by the bug in Dataflow that is losing state of PCollectionView on update.
* Kafka Reader in ingestion now has stable group id (across restarting jobs)

**Which issue(s) this PR fixes**:

Unblock #761 

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java [hunk 1]
```diff
@@ -137,38 +137,15 @@ public Job startJob(Job job) {
...
-      return job;
-    } catch (InvalidProtocolBufferException e) {
-      log.error(e.getMessage());
-      throw new IllegalArgumentException(
-          String.format(
-              "DataflowJobManager failed to UPDATE job with id '%s' because the job"
-                  + "has an invalid spec. Please check the FeatureSet, Source and Store specs. Actual error message: %s",
-              job.getId(), e.getMessage()));
-    }
+    abortJob(job.getExtId());
```
Comment (woop) [inline core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java:147]: Do we really need the `updateJob()` still if it can be replaced with `abortJob`?

File: core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java [hunk 2]
```diff
@@ -283,14 +260,17 @@ private ImportOptions getPipelineOptions(
     pipelineOptions.setJobName(jobName);
     pipelineOptions.setFilesToStage(
         detectClassPathResourcesToStage(DataflowRunner.class.getClassLoader()));
-
+    if (pipelineOptions.getTempLocation() == null && sink.getBigqueryConfig() != null) {
+      pipelineOptions.setTempLocation(sink.getBigqueryConfig().getStagingLocation());
```
Comment (woop) [inline core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java:264]: Is there a way to make this more generic (non-BQ specific)?

File: infra/scripts/test-templates/values-end-to-end-batch-dataflow.yaml [hunk 3]
```diff
@@ -30,7 +30,7 @@ feast-core:
             subnetwork: regions/$GCLOUD_REGION/subnetworks/$GCLOUD_SUBNET
             maxNumWorkers: 1
             autoscalingAlgorithm: THROUGHPUT_BASED
-            usePublicIps: false
+            usePublicIps: true
```
Comment (woop) [inline infra/scripts/test-templates/values-end-to-end-batch-dataflow.yaml:33]: How are we using public IPs here?

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java [hunk 4]
```diff
@@ -17,34 +17,30 @@
...
-  public GetTableDestination(String projectId, String datasetId) {
-    this.projectId = projectId;
-    this.datasetId = datasetId;
-  }
-
-  @Override
-  public TableDestination apply(ValueInSingleWindow<FeatureRow> input) {
-    String[] splitName = input.getValue().getFeatureSet().split("/");
+    if (splitName.length != 2) {
+      throw new RuntimeException("FeatureRow key is invalid. Must be project/feature-set");
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java:28]: If the `project` is not set then it should be configured to write to the `default` project. Meaning `project` is optional.

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java [hunk 5]
```diff
@@ -16,92 +16,201 @@
...
-            ParDo.of(
-                new DoFn<FeatureRowProto.FeatureRow, FeatureRowProto.FeatureRow>() {
-                  @ProcessElement
-                  public void processElement(ProcessContext context) {}
-                }));
-
     return WriteResult.in(input.getPipeline(), successfulInserts, failedElements);
   }
+
+  private PCollection<FeatureRow> mergeInputWithResult(
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java:142]: Would appreciate a few comments on this method.

**Comment by woop** [comment]
@pyalex would you mind pointing out all the code that we pulled up from Apache Beam? I would want to make sure that we give them proper attribution and that we don't put our license on their work.

File: core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java (hunk 1 above)
Comment (pyalex) [reply core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java:147]: I guess, there's nothing bad in having updateJob in the JobManager interface. There's still a chance that Dataflow update would be fixed in future and we'll switch back and also there could be other backends which support update w/o stop+start

File: core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java (hunk 2 above)
Comment (pyalex) [reply core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java:264]: we actually don't need this part at all, since now it's
`@NotBlank public String tempLocation;`

**Comment by pyalex** [comment]
> @pyalex would you mind pointing out all the code that we pulled up from Apache Beam? I would want to make sure that we give them proper attribution and that we don't put our license on their work.

sure, it's just one file https://github.com/feast-dev/feast/pull/757/files#diff-27d2cb5b830d488116fc8c364639b498

And I also added exclusion to spotless configuration https://github.com/feast-dev/feast/pull/757/files#diff-600376dffeb79835ede4a0b285078036R402 to avoid license addition. Not sure if there's better way to do it? @woop

**Comment by pyalex** [comment]
> I've left some comments, but they aren't fundamental changes. Just small nitpicks. Overall the PR looks good to me. One thing that I would ask for is improved JavaDoc coverage. This part of the code base can be hard to parse, so comments can only help the next person.

Added some details on big query writing flow as well

File: core/src/main/resources/application.yml [hunk 6]
```diff
@@ -38,7 +38,8 @@ feast:
     runners:
       - name: direct
         type: DirectRunner
-        options: {}
+        options:
+          tempLocation: gs://bucket/tempLocation
```
Comment (woop) [inline core/src/main/resources/application.yml:42]: Is this supposed to be under DataflowRunner? Or added here: https://github.com/feast-dev/feast/blob/master/protos/feast/core/Runner.proto#L33

File: docs/roadmap.md [hunk 7]
```diff
@@ -1,5 +1,24 @@
 # Roadmap
 
+## Feast 0.6 \(Feature Release\)
```
Comment (woop) [inline docs/roadmap.md:3]: Not sure where this came from, bad rebase?

File: core/src/main/resources/application.yml (hunk 6 above)
Comment (pyalex) [reply core/src/main/resources/application.yml:42]: it affects Direct as well, and was added here https://github.com/feast-dev/feast/pull/757/files#diff-ed1aa58bf58c59fe211c82b6df89182eR37

File: docs/roadmap.md (hunk 7 above)
Comment (pyalex) [reply docs/roadmap.md:3]: Apparently, I don't see any difference in my branch vs master on this file. So have no idea why it shows it here

File: core/src/main/resources/application.yml (hunk 6 above)
Comment (woop) [reply core/src/main/resources/application.yml:42]: Oops, I missed that. Thanks for pointing it out.

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryFeatureSink.java [hunk 8]
```diff
@@ -16,61 +16,99 @@
...
    * @param config {@link BigQueryConfig}
    * @param featureSetSpecs
-   * @return {@link BigQueryFeatureSink.Builder}
+   * @param bigQueryServices {@link BigQueryServices} - internal beam BigQuery wrapper
+   * @param bigQueryOptions {@link ValueProvider} value provider of BigQuery - from gcloud sdk
+   * @return
    */
-  public static FeatureSink fromConfig(
-      BigQueryConfig config, Map<String, FeatureSetProto.FeatureSetSpec> featureSetSpecs) {
+  static FeatureSink fromConfig(
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryFeatureSink.java:67]: Would it be possible to use the `Builder` directly instead of adding this `fromConfig`?

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryFeatureSink.java [hunk 9]
```diff
@@ -16,61 +16,99 @@
...
+  @Nullable
+  public abstract ValueProvider<BigQuery> getBigQueryOptions();
+
+  public abstract PCollection<KV<String, FeatureSetProto.FeatureSetSpec>> getFeatureSetSpecs();
 
   /**
-   * Initialize a {@link BigQueryFeatureSink.Builder} from a {@link BigQueryConfig}. This method
-   * initializes a {@link BigQuery} client with default options. Use the builder method to inject
-   * your own client.
+   * Mostly for internal use in tests Same as public version but allows to specify bigQuery clients
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryFeatureSink.java:58]: This is very nitpicky, but can you fix the casing/formatting of this sentence please.

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package feast.storage.connectors.bigquery.writer;
+
+import com.google.api.services.bigquery.model.TimePartitioning;
+import com.google.cloud.bigquery.DatasetId;
+import org.apache.beam.sdk.io.gcp.bigquery.TableDestination;
+
+public class BigQuerySinkHelpers {
+  /**
+   * Generating BQ table destination from dataset reference and featuresSet's project and name If
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java:25]: You are missing a period before the `If`

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+   * project is undefined "default" would be selected
+   *
+   * @param dataset {@link DatasetId} reference to bq project and dataset
+   * @param featureSetKey featureSet key in format project-name/set-name
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java:29]: `featureSetKey featureSet key in format project-name/set-name` > `featureSetRef Feature set reference with format <project>/<feature-set-name>`

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java [hunk 12]
```diff
@@ -16,92 +16,227 @@
...
     this.destination = destination;
+    this.schemas = schemas;
+  }
+
+  public BigQueryWrite withTriggeringFrequency(Duration triggeringFrequency) {
+    this.triggeringFrequency = triggeringFrequency;
+    return this;
+  }
+
+  public BigQueryWrite withTestServices(BigQueryServices testServices) {
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java:74]: Should this be called something like `withBigQueryService` in case it is used outside the testing context?

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java [hunk 13]
```diff
@@ -16,92 +16,227 @@
...
+              public void process(ProcessContext c) {
+                CoGbkResult result = c.element().getValue();
+                result
+                    .getAll(successTag)
+                    .forEach(success -> result.getAll(inputTag).forEach(c::output));
+              }
+            }));
+  }
+
+  private class FeatureDynamicDestinations extends DynamicDestinations<FeatureRow, String> {
```
Comment (woop) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java:216]: Can we get a JavaDoc here please.

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java (hunk 12 above)
Comment (pyalex) [reply storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java:74]: It's basically same method as here https://github.com/apache/beam/blob/b12a38aa2156c2652937669aeb2e46be15eca06c/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L2388
and there's no other (public) implementation except the test one, and it was done intentionally

File: storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+   */
+  public static TableDestination getTableDestination(DatasetId dataset, String featureSetKey) {
+    String[] splitName = featureSetKey.split("/");
+    String projectName, setName;
+
+    if (splitName.length == 2) {
+      projectName = splitName[0];
+      setName = splitName[1];
+    } else {
+      projectName = "default";
```
Comment (Yanson) [inline storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java:40]:
```suggestion
      projectName = Poject.DEFAULT_NAME;
```

Comment (pyalex) [reply storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java:40]: unfortunately Project model is in core, and there's reverse dependency core -> ingestion. maybe, it would be a good idea to move all models to `datatypes`, but I'm not sure it should be done within this PR. what about just local class constant for now?

Comment (woop) [reply storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQuerySinkHelpers.java]: > what about just local class constant for now?

This makes sense. I dont think it should be a part of this PR, but it's definitely a tech debt that should be cleared up.

**Comment by ches** [comment]
> * Dataflow JobManager stop+start job when specs is updated. This is caused by the bug in Dataflow that is losing state of PCollectionView on update.

Is there some further reference to what this about, that can be included in the description?

**Comment by ches** [comment]
> * Kafka Reader in ingestion now has stable group id (across restarting jobs)

Does this obviate #760 and close #646? It'd be ideal if this sort of change could be unbundled from a big PR like this—easier to track the status/closure of issues, easier to backport fixes versus enhancements, etc.

**Comment by woop** [comment]
> > * Kafka Reader in ingestion now has stable group id (across restarting jobs)
> 
> Does this obviate #760 and close #646? It'd be ideal if this sort of change could be unbundled from a big PR like this—easier to track the status/closure of issues, easier to backport fixes versus enhancements, etc.

I believe it does close those, will let @pyalex confirm. Yea, ideally those changes would not be bundled into one PR.

**Comment by pyalex** [comment]
@ches @woop Correct, issue with kafka consumer group should be solved and it was rather side effect not a main purpose of this PR

**Comment by ches** [comment]
> it was rather side effect not a main purpose of this PR

Understandable, glad for a fix getting mainlined either way!
